### PR TITLE
RFC: implement the Fernet format

### PR DESCRIPTION
**Pull request by timfeirg**
Specification: https://github.com/fernet/spec/blob/master/Spec.md

### CODE REVIEW

**Comment by leocavalcante** [comment]
Hi,
I don't know about Fernet format.
Can you give more info on a high-level? Is this kind of a JWT?

**Comment by timfeirg** [comment]
I learned about it from [this article](https://medium.com/coinmonks/if-youre-struggling-picking-a-crypto-suite-fernet-may-be-the-answer-95196c0fec4b).

It's like AES and HMAC combined into one protocol, at least that's my vague understanding.

**Comment by leocavalcante** [comment]
It seams interesting indeed. Thanks for contributing with the implementation.

**Comment by timfeirg** [comment]
dartanalyzer says there's type error with https://github.com/leocavalcante/encrypt/pull/66/files#diff-da91096b6b09f3c3e677dc9a05026a78R70, but I really don't understand what it's saying:

```
  error • Negation argument must have a static type of 'bool' at lib/src/algorithms/fernet.dart:70:10 • non_bool_negation_expression
```

**Comment by timfeirg** [comment]
this PR is complete except for the minor dartanalyzer mistake. @leocavalcante

**Comment by timfeirg** [comment]
In the Fernet spec, you can specify a ttl when decrypting messages, I've add the `ttl` parameter in [`Fernet.decrypt`](https://github.com/leocavalcante/encrypt/pull/66/files#diff-da91096b6b09f3c3e677dc9a05026a78R37), but there's no way to actually provide this argument since it's not specified in the declaration of `Algorithm.decrypt`.

I don't think ttl can be implemented since dart doesn't support varargs.

**Comment by timfeirg** [comment]
There's something funny going on with the [travis ci job](https://travis-ci.org/leocavalcante/encrypt/jobs/570689162), something seems to have gone wrong during `pub get`, but I can't reproduce on my computer, advice?

except from that, this PR is complete, request for comment @leocavalcante

File: lib/src/algorithms/fernet.dart [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+/// Wraps the Fernet Algorithm.
+class Fernet implements Algorithm {
+  final _maxClockSkew = 60;
+  final _randomGenerator = Random.secure();
+  Key _signKey;
+  Key _encryptionKey;
+  Clock _clock;
+
+  Fernet(Key b64key, {Clock clock}) {
+    final keyString = convert.utf8.decode(b64key.bytes);
```
Comment (leocavalcante) [inline lib/src/algorithms/fernet.dart]: I'm in doubt about this lines. Why all this conversion? Just `Key.bytes` isn't enough?

File: test/encrypt_test.dart [hunk 2]
```diff
@@ -9,6 +10,27 @@ void main() {
   const text = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit';
   final key = Key.fromUtf8('my32lengthsupersecretnooneknows1');
 
+  group('Fernet', () {
+    final currentDateTime =
+        DateTime.fromMillisecondsSinceEpoch(1565106118 * 1000);
+    final b64key = Key.fromUtf8(base64Url.encode(key.bytes));
```
Comment (leocavalcante) [inline test/encrypt_test.dart]: You are getting the UTF-8 key, converting its bytes to 64, then using a 64 bytes hash as UTF-8... Didn't get that part.

Comment (timfeirg) [reply test/encrypt_test.dart]: Fernet specified that the encryption key being a base64, utf8 encoded bytes.
I was translating code from [the python cryptography library](https://github.com/pyca/cryptography/blob/master/src/cryptography/fernet.py#L35).

File: lib/src/algorithms/fernet.dart (hunk 1 above)
Comment (timfeirg) [reply lib/src/algorithms/fernet.dart]: same reason as comment https://github.com/leocavalcante/encrypt/pull/66#discussion_r313688914.

`b64key.bytes` is a utf8 encoded, b64 encoded bytes, that's why I must first utf8 decode, and then b64 decode to get back the original `my32lengthsupersecretnooneknows1` key.

**Comment by timfeirg** [comment]
ping @leocavalcante

**Pull request merged**